src/cmd/metadata/contract.rs

```diff
@@ -0,0 +1,386 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// ink! is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+use core::{
+    fmt::{Display, Formatter, Result as DisplayResult, Write},
+    marker::PhantomData,
+};
+use semver::Version;
+use serde::{Serialize, Serializer};
+use serde_json::{Map, Value};
+use url::Url;
+
+const METADATA_VERSION: &str = "0.1.0";
+
+/// An entire ink! project for metadata file generation purposes.
+#[derive(Debug, Serialize)]
+pub struct ContractMetadata {
+    metadata_version: semver::Version,
+    source: InkProjectSource,
+    contract: InkProjectContract,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    user: Option<InkProjectUser>,
+    /// Raw JSON of the metadata generated by the ink! contract itself
+    #[serde(flatten)]
+    ink: Map<String, Value>,
+}
```

Comment:
Maybe you have misunderstood me in the call what I expect to live where or maybe there is generally a confusion on my side about this topic, however, what I meant was the following:

We want to have the full definitions (data structures such as `ContractMetadata`) to live in the `ink_metadata` crate. However, we do not want any of the code generation feature in that crate (e.g. what you did with `ToTokens` impls).
The `ink_metadata` should contain all the data structures and definitions for the whole ink! metadata definitions and should be able to serialize (and optionally deserialize) those to JSON and soon binary. That's it for `ink_metadata`.

What I expect to exist in `cargo-contract` is the code generation for what was formerly known by the `ink_abi` tool. So generation of code that is parameterized over what `cargo-contract` was able to query about the contract state. The generated code must be parameterized over the extended ink! metadata but should other than that be fairly straight forward. All the other definitions and algorithms about how ink! metadata is defined and how serialization is done should be in `ink_metadata` crate.

tl;dr: `cargo-contract` queries ink!'s `Cargo.toml` for extended metadata and generates the static code that itself generates the ink! metadata code with definitions from `ink_metadata`.

Reply:
> tl;dr: cargo-contract queries ink!'s Cargo.toml for extended metadata and generates the static code that itself generates the ink! metadata code with definitions from ink_metadata.

Indeed that's how I had it originally.

The main benefit of this approach is that the metadata generation is decoupled from the version of `ink!` for metadata generation (as long as the contract defines an `__ink_generate_metadata` extern). This means the tool could run metadata generation against different versions of `ink!` as it currently does. Also it solves the problem of brittle codegen for combining the extended metadata with the contract metadata.

However if this is not a requirement, and we want to release new `cargo-contract` major versions with `ink!` major versions, then I can move it all back and just do the codegen,

Reply:
See comment here also: https://github.com/paritytech/ink/issues/299#issuecomment-656750372